ouroboros-network/src/Ouroboros/Network/Protocol/TipSample/Codec.hs

```diff
@@ -0,0 +1,168 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE PolyKinds           #-}
+{-# LANGUAGE RankNTypes          #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeOperators       #-}
+
+module Ouroboros.Network.Protocol.TipSample.Codec
+  ( codecTipSample
+  , codecTipSampleId
+
+  , byteLimitsTipSample
+  , timeLimitsTipSample
+  ) where
+
+import           Control.Monad.Class.MonadST
+import           Control.Monad.Class.MonadTime
+
+import           Network.TypedProtocol.Pipelined (Nat (Succ, Zero), natToInt, unsafeIntToNat)
+
+import           Ouroboros.Network.Codec
+import           Ouroboros.Network.Driver.Limits
+import           Ouroboros.Network.Protocol.TipSample.Type
+import           Ouroboros.Network.Protocol.Limits
+
+import qualified Data.ByteString.Lazy as LBS
+
+import           Codec.CBOR.Decoding (decodeListLen, decodeWord)
+import qualified Codec.CBOR.Decoding as CBOR
+import           Codec.CBOR.Encoding (encodeListLen, encodeWord)
+import qualified Codec.CBOR.Encoding as CBOR
+import qualified Codec.CBOR.Read as CBOR
+import qualified Codec.Serialise as Serialise (Serialise (..))
+import           Text.Printf
+
+import           Unsafe.Coerce (unsafeCoerce)
+
+
+codecTipSample
+  :: forall m tip. MonadST m
+  => (tip -> CBOR.Encoding)
+  -> (forall s. CBOR.Decoder s tip)
+  -> Codec (TipSample tip)
+           CBOR.DeserialiseFailure m LBS.ByteString
+codecTipSample encodeTip decodeTip =
+    mkCodecCborLazyBS encode decode
+  where
+    encode :: forall (pr  :: PeerRole)
+                     (st  :: TipSample tip)
+                     (st' :: TipSample tip).
+              PeerHasAgency pr st
+           -> Message (TipSample tip) st st'
+           -> CBOR.Encoding
+    encode (ClientAgency TokIdle) (MsgFollowTip n slotNo) =
+      encodeListLen 2
+        <> encodeWord 1
```

Comment:
Why do you start encoding at 1 instead of at 0?

Reply:
Probably a left over; I'll fix this to start at `0`.